src-kmerge/KMerge/LoserTree.hs

```diff
@@ -35,69 +36,76 @@ placeholder = unsafeCoerce ()
 --
-newLoserTree :: forall a m. (PrimMonad m, Ord a) => [a] -> m (MutableLoserTree (PrimState m) a, Maybe a)
-newLoserTree [] = do
-    ids <- newPrimArray 0
-    arr <- newSmallArray 0 placeholder
-    sizeRef <- newPrimVar 0
-    holeRef <- newPrimVar 0
-    return $! (MLT sizeRef holeRef ids arr, Nothing)
-newLoserTree [x] = do
-    ids <- newPrimArray 0
-    arr <- newSmallArray 0 placeholder
-    sizeRef <- newPrimVar 0
-    holeRef <- newPrimVar 0
-    return $! (MLT sizeRef holeRef ids arr, Just x)
-newLoserTree xs0 = do
+newLoserTree :: forall a m. (PrimMonad m, Ord a) => NonEmpty a -> m (MutableLoserTree (PrimState m) a, a)
+newLoserTree zs0@(z :| zs) = do
     -- allocate array, we need one less than there are elements.
     -- one of the elements will be the winner.
-    ids <- newPrimArray  (len - 1)
-    setPrimArray ids 0 (len - 1) (-1)
-    arr <- newSmallArray (len - 1) placeholder
-
-    loop ids arr (len - 1) xs0
+    ids <- newPrimArray  len
+    arr <- newSmallArray len placeholder
+    case zs of
+      [] -> do
+        sizeRef <- newPrimVar 0
+        holeRef <- newPrimVar 0
+        return $! (MLT sizeRef holeRef ids arr, z)
+      _ -> do
+        setPrimArray ids 0 len (-1)
+        loop ids arr len zs0
   where
-    !len = length xs0
+    !len = length zs
 
-    loop :: MutablePrimArray (PrimState m) Int -> SmallMutableArray (PrimState m) a -> Int -> [a] -> m (MutableLoserTree (PrimState m) a, Maybe a)
-    loop !_   !_   !_   []     = error "should not happen"
-    loop  ids  arr  idx (x:xs) = do
+    loop :: MutablePrimArray (PrimState m) Int -> SmallMutableArray (PrimState m) a -> Int -> NonEmpty a -> m (MutableLoserTree (PrimState m) a, a)
+    loop  ids  arr  idx (x :| xs) = do
         sift ids arr (parentOf idx) (parentOf idx) x idx xs
 
-    sift :: MutablePrimArray (PrimState m) Int -> SmallMutableArray (PrimState m) a -> Int -> Int -> a -> Int -> [a] -> m (MutableLoserTree (PrimState m) a, Maybe a)
+    sift :: MutablePrimArray (PrimState m) Int -> SmallMutableArray (PrimState m) a -> Int -> Int -> a -> Int -> [a] -> m (MutableLoserTree (PrimState m) a, a)
     sift !ids !arr !idxX !j !x !idx0 xs = do
         !idxY <- readPrimArray ids j
         y     <- readSmallArray arr j
+        -- The algorithm has the invariant which states that,
+        -- if idxY < 0 then xs will be non-empty.
+        -- We check the invariant here and throw an exception
+        -- with a descriptive error message if it is violated.
         if idxY < 0
-        then do
+        then case xs of
+          -- We should never call 'loop' with an empty list,
+          [] -> error $ unlines
+            [ "Error in KMerge.LoserTree.newLoserTree"
+            , "Invariant violated:"
+            , "  EXPECTED    idxY < 0 ==> x : xs"
+            , "  OBSERVED    idxY < 0 and []"
+            , "  VALUE OF    idxY = " <> show idxY
+            ]
+          -- So only follow this logic branch if the list is non-empty!
```

Comment:
The comments here are not really needed anymore if we explain why `xs` can't be empty. 

The error message is also quite verbose, IMO without adding much information. The main point is that an invariant was violated, maybe with the info that `"xs = [], idxY = " show idxY`. The remaining information about the invariant can then be seen when looking at the code.